### PR TITLE
Ragged Buffer Factor support for ragged a2a approach

### PR DESCRIPTION
**Pull request by gobbleturk**
# Description

Add support for the users to control the size of the ragged buffer in the ragged a2a case (`use_ring_of_experts=False`). Support for the AG-RS case (`use_ring_of_experts=True`) will follow soon.

By default this feature will be off, and the ragged buffer will be worst case to ensure no dropping. Using this feature presents the risk of dropping. However this feature is still very useful:

* For many benchmarks, we can use random routing and set this factor to 1.0 or slightly larger
* Even real models should be fairly close to balanced, especially when using a small amount of EP (e.g. EP= 4 or 8).
* This feature can be used to show headroom for various operations to be handled as O(routed tokens) as opposed to O(worst case size) - e.g. this feature can be used to show how useful the ragged sort kernel is

Usage:
  * By default, `ragged_buffer_factor=-1.0`, which will not use the new future and stick to the old worst case size
  * Setting `ragged_buffer_factor=1.0` will use a ragged buffer that can hold perfectly balanced routing, but will drop otherwise
  * Setting`ragged_buffer_factor=2.0` will use a size twice as big, only drop if routing imbalance is 2x worse than balanced case, etc.

# Tests
Added tests to this PR for the ragged buffer factor calculation. Some manual tests run [here](https://docs.google.com/document/d/1t7Bb7aYFQ3BE6vnegCwBxFvwo7Ivb14Gd5-rjgU5d0E/edit?resourcekey=0-bLtIyKeq3S96dOEf9i343Q&tab=t.0)

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/configs/base.yml [hunk 1]
```diff
@@ -195,6 +195,11 @@ num_experts_per_tok: 1
 megablox: true
 sparse_matmul: true
 capacity_factor: -1.0 # a factor to decide expert capacity for token dropping, and no dropping by default
+ragged_buffer_factor: -1.0 # a factor to determine the size of the ragged buffer for routed MoE activations.
+# By default (-1), the routed buffer will be worst case size to ensure no dropping.
+# When set to 1.0 this buffer if set to the size assuming perfectly balanced. If the routing dictates
+# a size larger than this then tokens will be dropped.
+# In general if ragged_buffer_factor>0, the ragged_buffer_size is is balanced_size * ragged_buffer_factor.
```
Comment (RissyRan) [inline src/maxtext/configs/base.yml]: nit `ragged_buffer_factor > 0` & remove duplicate `is`

File: src/maxtext/configs/types.py [hunk 2]
```diff
@@ -2095,6 +2090,12 @@ def load_model_specific_defaults(cls, values: dict[str, Any]) -> dict[str, Any]:
     """This method is a no-op because `pyconfig` handles model-specific config loading."""
     return values
 
+  def validate_ragged_buffer_factor(self):
+    if self.ragged_buffer_factor == -1.0:
```
Comment (RissyRan) [inline src/maxtext/configs/types.py]: Do we need this check?

File: src/maxtext/configs/types.py [hunk 3]
```diff
@@ -2095,6 +2090,12 @@ def load_model_specific_defaults(cls, values: dict[str, Any]) -> dict[str, Any]:
...
+      return # Nothing to validate if not using ragged buffer factor
+    if self.use_ring_of_experts:
```
Comment (NuojCheng) [inline src/maxtext/configs/types.py:2092]: I would add a warning massge here indicating the tokens can be dropped when ragged_buffer_factor is on.

File: tests/unit/moe_test.py [hunk 4]
```diff
@@ -1062,6 +1062,59 @@ def test_get_all_to_all_params_unsharded_batch(self):
...
+    num_experts_per_tok = 8
+    global_experts = 256
+    ep_degree = 128    
+    
+    expected_ragged_buffer = 1048576 # (32768) * (global_exp / top_k)
+    actual_ragged_buffer = moe.RoutedMoE.get_ragged_buffer_size(local_batch, ep_degree, global_experts, num_experts_per_tok, ragged_buffer_factor)
+    self.assertEqual(expected_ragged_buffer, actual_ragged_buffer)
+    self.assertEqual(True, True)
+
+
```
Comment (RissyRan) [inline tests/unit/moe_test.py:1118]: nit: extra lines

File: src/maxtext/layers/moe.py [hunk 5]
```diff
@@ -1148,59 +1177,40 @@ def wrapper(x, logits, pre_bias_logits, w0, w1, wo, w0_bias, w1_bias, wo_bias, r
...
-            )
+          all_shards_group_sizes = jax.lax.all_gather(reshaped_group_sizes, axis_name=self._expert_parallelism_name)
+          input_offsets, send_sizes, output_offsets, recv_sizes = RoutedMoE.get_all_to_all_params(
+              all_shards_group_sizes,
+              expert_shard_id,
+              num_expert_parallelism,
+          )
+
+          max_local_experts_per_tok = min(local_expert_size, self.config.num_experts_per_tok)
+          # def get_ragged_buffer_size(local_batch, ep_degree, global_experts, num_experts_per_tok, ragged_buffer_factor)
```
Comment (NuojCheng) [inline src/maxtext/layers/moe.py]: remove this line?

File: tests/unit/moe_test.py [hunk 6]
```diff
@@ -1062,6 +1062,59 @@ def test_get_all_to_all_params_unsharded_batch(self):
...
+    ragged_buffer_factor = -1.0 # Not using ragged_buffer_factor
+    local_batch = 32768
+    num_experts_per_tok = 8
+    global_experts = 256
+    ep_degree = 4    
+    
+    expected_ragged_buffer = 131072 # local_batch * ep_degree
+    actual_ragged_buffer = moe.RoutedMoE.get_ragged_buffer_size(local_batch, ep_degree, global_experts, num_experts_per_tok, ragged_buffer_factor)
+    self.assertEqual(expected_ragged_buffer, actual_ragged_buffer)
+    self.assertEqual(True, True)
```
Comment (NuojCheng) [inline tests/unit/moe_test.py]: ???? Gemini did that?

File: tests/unit/moe_test.py [hunk 7]
```diff
@@ -1062,6 +1062,59 @@ def test_get_all_to_all_params_unsharded_batch(self):
...
+    ragged_buffer_factor = -1.0 # Not using ragged_buffer_factor
+    local_batch = 32768
+    num_experts_per_tok = 8
+    global_experts = 256
+    ep_degree = 128    
+    
+    expected_ragged_buffer = 1048576 # (32768) * (global_exp / top_k)
+    actual_ragged_buffer = moe.RoutedMoE.get_ragged_buffer_size(local_batch, ep_degree, global_experts, num_experts_per_tok, ragged_buffer_factor)
+    self.assertEqual(expected_ragged_buffer, actual_ragged_buffer)
+    self.assertEqual(True, True)
```
Comment (NuojCheng) [inline tests/unit/moe_test.py]: do we really need 4 tests protecting it? seem dumb lol

**Review by NuojCheng — APPROVED** [review]

File: src/maxtext/layers/moe.py (hunk 5 above)
Comment (gobbleturk) [reply src/maxtext/layers/moe.py]: Done

File: src/maxtext/configs/types.py [hunk 8]
```diff
@@ -2095,6 +2090,12 @@ def load_model_specific_defaults(cls, values: dict[str, Any]) -> dict[str, Any]:
     """This method is a no-op because `pyconfig` handles model-specific config loading."""
     return values
 
+  def validate_ragged_buffer_factor(self):
+    if self.ragged_buffer_factor <= 0:
```
Comment (RissyRan) [inline src/maxtext/configs/types.py:2090]: Do we need this?

File: src/maxtext/configs/types.py (hunk 3 above)
Comment (RissyRan) [reply src/maxtext/configs/types.py:2092]: Hmm... I think it's better to throw early error to users and indicate what they are doing.

**Review by RissyRan — COMMENTED** [review]
LGTM! Thanks! Could you help add this config into https://github.com/AI-Hypercomputer/maxtext/blob/083293fc373bb26dece7d81901a13f6e9c3e584f/docs/reference/core_concepts/moe_configuration.md?

**Comment by RissyRan** [comment]
Gemini review doesn't work with PR conflicts. Recommend to rebase and re-trigger the `gemini-review` label :)

File: tests/unit/moe_test.py (hunk 4 above)
Comment (gobbleturk) [reply tests/unit/moe_test.py:1118]: Done.

File: src/maxtext/configs/base.yml (hunk 1 above)
Comment (gobbleturk) [reply src/maxtext/configs/base.yml]: Done.
